include/ppx/knob.h

```diff
@@ -0,0 +1,284 @@
+        mMinValue = minValue;
+        mMaxValue = maxValue;
+        mCallback = callback;
+        SetValue(defaultValue, true);
+    }
+
+    void Draw() override;
+    void ResetToDefault() override { SetValue(mDefaultValue); }
+
+    // FlagText format:
+    // --flag_name <mMinValue~mMaxValue> : flag description
+    std::string FlagText() const override;
+
+    // Updates mValue and mDefaultValue, expected flag format:
+    // --flag_name <int>
+    Result UpdateFromFlag(const CliOptions& opts) override;
+
+    int GetValue() const { return mValue; }
+
+    // Used for when mValue needs to be updated outside of UI
+    // Returns success if newVal is valid, fails otherwise
+    // If newVal is valid:
+    // - If updateDefault, mDefaultValue is updated
+    // - If newVal is different from mValue:
+    //   - mValue is updated
+    //   - mCallback is triggered
+    Result SetValue(int newVal, bool updateDefault = false);
+
+private:
+    int mValue;
+    int mDefaultValue;
+
+    // mValue will be clamped to the mMinValue to mMaxValue range, inclusive
+    int mMinValue;
+    int mMaxValue;
+
+    // mCallback will be called on the new int value every time the slider is deselected in the UI, or updated with SetIntValue()
+    std::function<void(int)> mCallback;
+};
+
+// KnobStrDropdown is an int knob that will be displayed as a dropdown in the UI
+// The knob stores the index of a selected choice from a list of allowed strings
+class KnobStrDropdown
+    : public Knob
+{
+public:
+    KnobStrDropdown(const std::string& displayName, const std::string& flagName, const std::string& flagDesc, int defaultIndex, const std::vector<std::string>& choices, std::function<void(int)> callback = NULL)
+        : Knob(displayName, flagName, flagDesc)
+    {
+        if (choices.size() < 1) {
+            PPX_LOG_FATAL(flagName << " KnobStrDropdown: too few allowed choices");
+        }
+        mChoices  = choices;
+        mCallback = callback;
+        auto res  = SetIndex(defaultIndex, true);
+        if (res != SUCCESS)
+            PPX_LOG_FATAL(flagName << "cannot be initialized");
+    }
+
+    void Draw() override;
+    void ResetToDefault() override { SetIndex(mDefaultIndex); }
+
+    // FlagText format:
+    // --flag_name <"choice1"|"choice2"|...> : flag description
+    std::string FlagText() const override;
+
+    // Updates mIndex and mDefaultIndex, expected flag format:
+    // --flag_name <str>
+    Result UpdateFromFlag(const CliOptions& opts) override;
+
+    int         GetIndex() const { return mIndex; }
+    std::string GetStr() const { return mChoices[mIndex]; }
+
+    // Used for when mIndex needs to be updated outside of UI
+    // Returns success if newI is valid, fails otherwise
+    // If newI is valid:
+    // - If updateDefault, mDefaultIndex is updated
+    // - If newI is different from mIndex:
+    //   - mIndex is updated
+    //   - mCallback is triggered
+    Result SetIndex(int newI, bool updateDefault = false);
```

Comment:
Templated knob means those 2 versions could be removed.
Also they should probably cease to exist if you follow my recommendation to not set the value later in the knob lifecycle.
ImGUI will drive the value from the Draw() function. Seems like there is no need to drive this value from the outside then.

Reply:
Ok, got it, understand why the Set() is needed then.